### ISSUE TITLE
[BUG] Process.env not defined in mounted component

### ISSUE THREAD

**Issue by divyanshu-rawat** [microsoft/playwright#21187]
**Context:**
- Playwright Version:  `"@playwright/experimental-ct-react": "^1.31.0",`
- Operating System: [ Mac]
- Node.js version: v16.18.0
- Browser: [e.g. All, Chromium]

**Code Snippet**
Here is how my playwright-ct.config.ts looks like.

```js

import { resolve } from 'path';

import { defineConfig, devices } from '@playwright/experimental-ct-react';
import dotenv from 'dotenv';

dotenv.config({ path: './env/.env.test', override: true });

export default defineConfig({
  testDir: './playwright/component',
  snapshotDir: './playwright/__snapshots__',
  timeout: 10 * 1000,
  fullyParallel: true,
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 2 : 0,
  workers: process.env.CI ? 1 : undefined,
  reporter: 'html',
  use: {
    trace: 'on-first-retry',
    baseURL: 'https://localhost',
    ctPort: 3100,
    ctViteConfig: {
      resolve: {
        alias: {
          '@utils': resolve(__dirname, '../ui/src/utils'),
          'process': 'process/browser'
        },
        envDir: './env/.env.test',
        define: {
          'process.env': process.env
        }
      }
    }
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] }
    }
  ]
});
```

My Directory structure.

```
./playwright
 - component
     ./ MyComponent.spec.tsx
     ./ MyComponent.tsx
     ./ index.html
 - fixtures
 - index.ts

```

Here is the test case that mounts the My component,  `./MyComponent.spec.tsx`

```js
import React from 'react';

import { test, expect } from '@playwright/experimental-ct-react';

import { MyComponent } from './myComponent';

test.beforeAll(async () => {
  process.env.TEST_ENV_VAR = 'test_value';
});

test('mount My Component', async ({ mount, page }) => {
  // console.log(process.env) -->> works here it is accessible.
  await page.pause();
  const component = await mount(<MyComponent />);
});
```

Here is how my component looks like. `./MyComponent.tsx`

```js

import { useAnalytics } from '@hooks/useAnalytics';

export const MyComponent = () => {
  const { shareEvent } = useAnalytics();
 // process.env is not defined inside my component causing useAnalytics() hook to not work properly as it is using.
  // environment variables to load the script.
  // console.log('process.env', process.env); ->> this is undefined, doesn’t work here.

  return (
    <>
      <button onClick={() => shareEvent("PDP")}>shareEvent</button>
    </>
  );
};

```

**Describe the bug**

The component that needs to be mounted using playwright component testing had `process.env` undefined, In `MyComponent.spec.tsx` process.env work fine, wherever in the component `./MyComponent.tsx` it doesn’t work it is undefined.

In Summary: **Environment variables are accessible in the test that is used to mount the component for component testing but undefined in component itself.**

**Comment by yury-s** [comment]
The test runs in the Node.js process while the component is created inside a real browser in a separate process. `process` is not a part of Web API, this is why you can't access it from your component. You can use [`page.evaluate()`](https://playwright.dev/docs/api/class-page#page-evaluate) to run javascript code in the browser context:

```js
test.beforeEach(async ({ page }) => {
  await page.evaluate(() => (window as any).TEST_ENV_VAR = 'test_value');
});

test('should work', async ({ mount }) => {
  const component = await mount(<App />);
  await expect(component).toContainText('Test');
});
```
and then in your component:
```ts
export const MyComponent = () => {
  const { shareEvent } = useAnalytics();

  console.log('TEST_ENV_VAR = ' + (window as any).TEST_ENV_VAR);

  return (
    <>
      <button onClick={() => shareEvent("PDP")}>shareEvent</button>
    </>
  );
};
```

Does this work for you?

**Comment by divyanshu-rawat** [comment]
@yury-s yes thanks for your response, it is helpful :)

So, In summary, my test component renders a hook `useAnalytics()` `import { useAnalytics } from '@hooks/useAnalytics';` that is being used to load an analytics script into the browser, and to load that script the URL comes from the environment variable  ( process.env.ANALYTICS_URL ).

I consider it a common scenario when you want to test components using Playwright and there should be some solution to this problem. I am using the next.js framework to build the application.

**Comment by yury-s** [comment]
Vite exposes environment variables on `import.meta.env.` see [their doc](https://vitejs.dev/guide/env-and-mode.html) for more details. Note that by default they only copy variables prefixed with `VITE_`. The following code worked fine for me:

```tsx
import Button from './components/Button';

export default function App() {
  console.log('VITE_TEST_ENV_VAR = ' + import.meta.env.VITE_TEST_ENV_VAR);
  return <>
      <Button title='Test'></Button>
  </>
}
```
when running with `VITE_TEST_ENV_VAR=bar npm run test-ct`. This [comment](https://github.com/microsoft/playwright/issues/18076#issuecomment-1282777975) may also be helpful.

Closing this issue as how the env variables are handled and baked into the app is part of Vite's logic. Please refer to their [documentation](https://vitejs.dev/guide/env-and-mode.html#env-variables) for more details. Feel free to file a new issue if it still doesn't work.

**Comment by ozaytsev86** [comment]
Hi  everyone, thanks for the answer @yury-s . Is there a way do that globally instead of doing it in each test file?
